### PR TITLE
Python blackoil simulator

### PR DESCRIPTION
**Pull request by jmsargado**
Python bindings for blackoil simulator using pybind11 (very much still a work in progress). More stuff to follow later.

### CODE REVIEW

**Comment by joakim-hove** [comment]
The changes in the commit [Skip processing of filename ...](https://github.com/OPM/opm-simulators/pull/2127/commits/08d38fe492bfe23a8ee3afc63c1146d820271f1b) is independent of Python or not and will affect flow usage - I think that should be extracted to a separate pull request.

**Comment by joakim-hove** [comment]
The bindings part in the last commit seems to create a full new simulator implementation; I would *really* have hoped that it should be possible to bind to one of the existing simulators[1] without code duplication.

[1]: Maybe some changes must be made to the existing simulator code, then that should be done first - in separate small pull requests.

File: python/simulators/simulators.cpp [hunk 1]
```diff
@@ -226,6 +233,16 @@ class BlackOilSimulator
         delete[] argv_;
     }
 
+    void setupCmdLineArgs()
```
Comment (joakim-hove) [inline python/simulators/simulators.cpp]: I do not understand what this function is needed/used for. Maybe it can be removed entirely?

**Comment by joakim-hove** [comment]
> I've modified the python simulator constructor to accept either the deck filename
'
That seems like a good idea - but I can not see the overload which uses a filename?

**Comment by jmsargado** [comment]
The first constructor implementation accepts a filename, copies it to argv_[1] and then let's the rest of the internal code in the run method do the work. I've put back the code from flow.cpp that generates the deck, eclipse state and other objects given the deck filename. This is then simply skipped when such objects already exist (which is the case when the other constructor is called).

**Comment by jmsargado** [comment]
The setupCmdLineArgs() is just there to setup the internally stored argc and argv that is passed to other functions the need them. Originally the code was part of the constructor but as there's now more than one constructor, I felt it would be easier to have it separate to avoid having to write the same thing twice.

**Comment by joakim-hove** [comment]
I will continue reviewing this tomorrow, it would be good if you could rebase it in the meantime so the diff is a bit more manageable.

**Comment by joakim-hove** [comment]
I have extracted the pure C++ change in a separate PR: https://github.com/OPM/opm-simulators/pull/2151

**Comment by joakim-hove** [comment]
> it would be good if you could rebase it in the meantime so the diff is a bit more manageable.

You really have to rebase this:
```
bash% git fetch upstream
bash% git checkout python_blackoil_simulator
bash% git rebase upstream/master
bash% git push -f origin python_blackoil_simulator
```
it is too cumbersome to review when all the changes from #2126 are also included in the diff.

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
I have created this : opm/opm-common#1545 to ensure that the Python extension is indeed built by jenkins.

**Comment by joakim-hove** [comment]
Changes are required to `python/simulators.cpp`:

1. The code makes use of `boost::filesystem` - that is no longer available in opm code, you should use the `std::filesystem / std::experimental::filesystem` wrapper available in opm-common:
```
#include <opm/common/utility/FileSystem.hpp>


using fs = Opm::filesystem;
...

``` 
2. It looks like there is much duplicated code between `python/simulators.cpp` and `flow.cpp` - that must be resolved in some way.

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by bska** [comment]
Closing this now since the patch no longer applies.

**Pull request closed without merging**